### PR TITLE
added semaphore

### PR DESCRIPTION
**Pull request by gsbelarus**

### CODE REVIEW

File: src/Semaphore.ts [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+      this._queue.shift()?.();
+    } else {
+      this._permits += 1;
+    }
+
+    if (this._permits > this._maxAllowed) {
+      console.error('Release() has been called without previous acquire!');
+    }
+  }
+}
```
Comment (gsbelarus) [inline src/Semaphore.ts:46]: Here's a concise review of the code provided: 

1. **Purpose**: The `Semaphore` class is designed to manage access to a finite number of resources or permits.

2. **Type Definition**: The `Unlock` type is an alias for a function that returns void, likely intended as a callback signature.

3. **Constructors and Properties**:
   - Defaulting `maxAllowed` to 1 in the constructor is sensible for creating mutex (binary semaphore) behavior.
   - Good use of getters for `permits`, `queueLength`, and `acquired`.

4. **Acquire Method**:
   - Uses an asynchronous pattern correctly, returning immediately if a permit is available or returning a promise that will resolve when the permit can be acquired.
   - If `this._permits` is exactly 0, the acquire method will neither decrement `_permits` nor push a new promise to `_queue`. It should have an `else if` condition to handle this situation correctly.

5. **Release Method**:
   - Correctly prioritizes releasing a waiting task in the queue over incrementing `_permits`.
   - However, releasing a permit from an empty semaphore should not increase `_permits` beyond `_maxAllowed`. This can happen here since the check only occurs after potentially adding to `_permits`. It is better to do the check before increasing `_permits`.

6. **Error Handling**:
   - The error message in `release()` can warn developers if there's a logical fault causing premature release calls.

7. **Concurrency Safety**: 
   - One concurrency issue arises when there are multiple async calls to the `acquire` and `release` methods. The code does not guard against race conditions when modifying `_permits` or altering the `_queue`.

8. **Testing and Edge Cases**:
   - It would be good to test edge cases where permits are released without being acquired first and acquiring permits with simultaneous calls that exceed the maximum allowed.

9. **Code Style**: 
   - Code style seems consistent and clear. Naming conventions are followed, enhancing readability.

10. **Additional Suggested Improvements**:
    - Include validation for `maxAllowed` input in the constructor to ensure it's positive.
    - Consider using a `while` loop in `release` to handle scenarios where multiple promises can be resolved at once (if that's desired behavior).
    - Adding comments or documentation comments could help future maintainers understand the intent and usage of each part of the Semaphore class.

In summary, the code appears well-structured for a basic semaphore implementation but requires fixes to ensure proper synchronization among concurrent operations and prevent the semaphore state from becoming invalid. Additional safeguards and tests would strengthen the reliability of the implementation.
